crypto: Deprecate createCipher for createCipheriv

iangcarroll opened the pull request:
This is the start of an implementation of @cbarcenas's [proposal](https://github.com/nodejs/node/pull/13821#issuecomment-309900506) and an evolution of #13821, except `createCipher` ends up being deprecated. I believe the poor KDF in `createCipher` is also harmful, as it allows arbitrary length keys but only puts them through a round of MD5, hence the removal. 

* `crypto.createCipher` is deprecated entirely. Initially, this pull request made `createCipher` error when passed an IV-dependent cipher, but this has been removed.
* `crypto.createCipheriv` now supports a `null` IV for algorithms without an IV.

This is my first Node PR, so please let me know if I've done anything wrong.

- [x] Raise a deprecation warning when `createCipher` is called
- [x] Implement `null` IV support into `createCipheriv` for use with ECB
- [x] Expose the old and insecure KDF in a new API method

```js
> let cipher = crypto.createCipher('aes-128-gcm', new Buffer(crypto.randomBytes(16)));
Error: crypto.createCipher() is no longer supported with ciphers that require initialization vectors. Generate a random IV and pass it to crypto.createCipheriv().

> let cipher = crypto.createCipheriv('aes-128-ecb', new Buffer(crypto.randomBytes(16)), null); cipher.update('abc', 'utf-8'); cipher.final('hex');
'e5791e621970d682bb0e353b7bbc6fce'
```

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
crypto, doc

jasnell commented on src/node_crypto.cc:
```diff
@@ -3349,11 +3348,19 @@ void CipherBase::Init(const char* cipher_type,
+
+  if (encrypt && iv_len != 0) {
+    return env()->ThrowError("crypto.createCipher() is no longer supported with ciphers that require initialization vectors. Generate a random IV and pass it to crypto.createCipheriv().");
+  }
```
No comments on whether this is something we should do (ping @nodejs/crypto) ... but this is not the correct way of doing it. There are two kinds of deprecations we use: documentation and runtime. A runtime deprecation uses either the `util.deprecate()` or `process.emitWarning('...', 'DeprecationWarning')` API to emit a warning at runtime, but the code still continues to operate as it did in every other respect.

iangcarroll replied:
My PR intends to immediately prohibit the use of this API for ciphers that require an IV. The deprecation of this API entirely hasn’t been done yet. I’m guessing that would go in lib/crypto.js?

jasnell replied:
Yep, and that's the point: Our deprecation policy *requires* that APIs go through a proper deprecation cycle so this PR would not be able to land as is. The first step is to do a docs or runtime deprecation as a semver-major, which would go into Node.js 9.0.0.

bnoordhuis replied:
We do have provisions for making exceptions for security-related issues but as this is more about protecting users against themselves, it's up for debate whether those provisions apply.

I'm personally leaning towards 'acceptable in node 9' (raising an exception, that is) but I won't hold it against anyone who feels differently.

shigeki replied:
This prohibits users from using not only counter modes but also cbc modes.  I'm not sure if the cbc mode has a security issue in this API.

jasnell commented on src/node_crypto.cc:
```diff
@@ -3443,14 +3450,22 @@ void CipherBase::InitIv(const FunctionCallbackInfo<Value>& args) {
+
+  if (!Buffer::HasInstance(args[2]) && !args[2]->IsNull()) {
+    return env->ThrowTypeError("IV must be a buffer or null");
+  }
```
why this change?

iangcarroll replied:
Because this API will now accept null as a valid IV.

iangcarroll replied:
(This means you don’t need to pass an empty Buffer for ciphers where there is no IV)

jasnell replied:
The better approach here then, would be to do the type checking at the js layer and use the internal/errors API to generate a proper `ERR_INVALID_ARG_TYPE` error.

tniessen replied:
Is this about `createCipheriv`? Why would any sane person come up with the idea to use a function `createCipheriv` for ciphers where there is no IV?

tniessen commented:
This might be a start, but I don't think we can land this without the other suggested changes. Summary of those changes:

- Deprecate `createCipher` support for all ciphers which require an `IV`.
- Add `generateLegacyKey` and `generateLegacyIV`.

Without the second point implemented, I would not want to land the first one.

> crypto.createCipher will no longer function for ciphers that require an initialization vector, and is deprecated entirely.

Why would we want to deprecate it as a whole? It is perfectly fine to use it with ciphers that do not require an IV.

Additionally, we should consider @stouset's https://github.com/nodejs/node/pull/13821#issuecomment-309906223, but for reasons discussed in #13821, we might want to give that function a new name.

iangcarroll commented:
I didn't mean for this to be immediately landed, just for discussion to happen over what's done so far.

> It is perfectly fine to use it with ciphers that do not require an IV.

Not sure I agree. First, the fact it takes a `password` and not (always) a key buffer is non-standard. But more importantly it derives a key very insecurely. So the two options are essentially to fix it, to the extent we can (i.e. we can't easily add a salt), or pass this duty to the user and force a proper key.

We could change `createCipher`'s signature to require a proper-length Buffer instead of removing it. That does sound more reasonable than pushing it to `createCipheriv`.

I guess this all depends on whether or not these APIs are supposed to be low or high level, as @cbarcenas has mentioned.

@stouset's proposal probably needs its own PR.

shigeki commented:
I was conservative to deprecate this API because it is very old API and I could not imagine how much it is affected like Buffer API.

>  I believe the poor KDF in createCipher is also harmful, as it allows arbitrary length keys but only puts them through a round of MD5, hence the removal.

I agree this. It is a harmful to have md5 KDF at this time. If we can really proceed to make deprecation, I think we had better to deprecate entire `createCipher()`.

tniessen commented:
I see your point, the current key generation is not too secure... I actually expected a way to change the signature such that users could either specify a password *or* a key, but I don't see such a solution right now. We might actually have to deprecate the whole function, which is terrible, as `createCipher` was such a nice name for the function. (Plus, I don't like having to call a function `createCipheriv` for ciphers without an IV). So I would still be happy if we found another solution.

stouset commented:
@tniessen Creating ciphers without an IV is inherently dangerous, and in my opinion should not be the easiest way to invoke the library. It's fine for this to be considered a low-level API, but without a very widely-promoted high-level one, a module named `crypto` is going to be reached for by pretty much anyone searching for how to do it in Node.

ECB is the only mode I can think of that doesn't require instantiation with an IV. Given that, an API like `createUnsafeRawBlockCipher` is in my eyes significantly better for a function that instantiates ECB, in that it steers people away who have no business touching raw block ciphers.

iangcarroll commented:
I've pushed the deprecation warning for `createCipher` and `Cipher`, along with the documentation for the deprecation (not yet for the function). I also made the `createCipheriv` call easier for ECB mode, so explicitly providing null is no longer required.

One thing I saw is the `options` argument to `createCipher` and `createCipheriv`, which is undocumented but used in a call to `LazyTransform`. Can this be removed or is it used somewhere?

Also, we could think about using `crypto.Cipher()` as an alias/replacement for `createCipheriv`, seeing how its usage has not been documented (though it currently points to `createCipher`).

I still need to add the legacy methods, more documentation, and tests.

iangcarroll commented:
`generateLegacyKey` and `generateLegacyIV` have been implemented.

**Old:**
```js
> let cipher = crypto.createCipher('aes-128-ecb', 'password');
undefined
> cipher.update('hello');
<Buffer >
> cipher.final('base64');
'LJ6pZaGs8G7A/W5gW+txkw=='
```

**New:**
```js
> var cipher = crypto.createCipheriv("aes-128-ecb", crypto.generateLegacyKey("aes-128-ecb", "password"));
undefined
> cipher.update("hello");
<Buffer >
> cipher.final("base64");
'LJ6pZaGs8G7A/W5gW+txkw=='
```

iangcarroll commented:
All tests/linters now pass on my machine. Can someone trigger a CI build? Not sure how to go about that.

tniessen commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9055/

BridgeAR commented:
This needs a rebase and it seems like not all tests passed.

iangcarroll commented:
I can finish this up soon. I believe there is a bug (with the tests?) when FIPS mode is used.

BridgeAR commented:
Ping @iangcarroll

jasnell commented on doc/api/deprecations.md:
```diff
@@ -634,6 +634,19 @@ Type: Runtime
+
+Type: End-of-Life
+
+[`crypto.createCipher()`][] generates keys from strings in an insecure manner, and, when used with a cipher that utilizes an initialization vector, will dangerously re-use initialization vectors. As such, it is immediately marked as End-of-Life when used with ciphers that require initialization vectors, and will be fully removed in a later version.
```
Please line wrap at <= 80 chars

jasnell commented on doc/api/deprecations.md:
```diff
@@ -634,6 +634,19 @@ Type: Runtime
+<a id="DEP00XX"></a>
+### DEP00XX: crypto.createCipher()
+
+Type: End-of-Life
```
This should not be automatically moved to `End-of-Life`. This needs to go through a proper deprecation cycle with either a Documentation-only or Runtime stage.

jasnell commented on doc/api/deprecations.md:
```diff
@@ -634,6 +634,19 @@ Type: Runtime
+
+[`crypto.createCipher()`][] generates keys from strings in an insecure manner, and, when used with a cipher that utilizes an initialization vector, will dangerously re-use initialization vectors. As such, it is immediately marked as End-of-Life when used with ciphers that require initialization vectors, and will be fully removed in a later version.
+
+[`crypto.createCipheriv()`][] should be used in place of [`crypto.createCipher()`][]. Since [`crypto.createCipheriv()`][] will no longer attempt to derive a proper encryption key from a string, you must use a key-derivation function such as [`crypto.pbkdf2()`][] to obtain a valid key if you normally supply a string to [`crypto.createCipher()`][].
```
Please avoid using the informal pronoun `you` in the documentation.

iangcarroll commented:
Can someone please re-run the CI build? I can't reproduce the failing build locally.

Is there any sort of process for deciding the plan for deprecating this? Or is there no hope of directly EOLing it?

tniessen commented:
Here is another CI: https://ci.nodejs.org/job/node-test-pull-request/10099/
(This CI should not be considered when deciding whether to land this PR as it was not rebased on master.)

BridgeAR commented:
@iangcarroll this needs a rebase and the comments should be addressed.

About going directly to EOL - I understand your concerns but even if this is only a run time deprecation it should already help a lot. For further informations about our deprecation cycles, please have a look at the [guide](https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#deprecations).

BridgeAR commented:
@iangcarroll do you still want to pursue this PR?

iangcarroll commented:
Yeah, I have local fixes for the CI failures, there's just one more bug I need to fix with current (mis)uses of `createCipheriv`, particularly that these functions return different things, which I did not expect and don't actually understand at the moment.

```js
> let encrypt = crypto.createCipheriv('BF-ECB', 'SomeRandomBlahz0c5GZVnR', '');
undefined
> encrypt.update('Hello World!', 'ascii', 'hex');
'6d385f424aab0cfb'
```

```js
> let encrypt = crypto.createCipher('BF-ECB', 'SomeRandomBlahz0c5GZVnR');
undefined
> encrypt.update('Hello World!', 'ascii', 'hex');
'b2941213258d679b'
```

bnoordhuis commented:
That's because `crypto.createCipher()` generates a key with `EVP_BytesToKey(EVP_md5())`.  That is, these two produce identical output:
```
> crypto.createCipher('bf-ecb', 'pass').update('11223344', 'utf8', 'hex')
'93b5c44c33c68cd1'

> var pass = crypto.createHash('md5').update('pass').digest();
> crypto.createCipheriv('bf-ecb', pass, '').update('11223344', 'utf8', 'hex')
'93b5c44c33c68cd1'
```

iangcarroll commented:
I have changed the deprecation so it is now a runtime warning for all calls to `createCipher`, and added documentation for the changes. Let me know how it looks -- still a bit new to this so might be missing some stuff.

tniessen commented on benchmark/crypto/cipher-stream.js:
```diff
@@ -33,8 +33,11 @@ function main(conf) {
+  const iv = crypto.generateLegacyIV(conf.cipher, alice_secret);
+
+  var alice_cipher = crypto.createCipheriv(conf.cipher, key, iv);
+  var bob_cipher = crypto.createDecipheriv(conf.cipher, key, iv);
```
Use `const` instead of `var`.

tniessen commented on doc/api/crypto.md:
```diff
@@ -1214,6 +1218,44 @@ The `key` is the raw key used by the `algorithm` and `iv` is an
+- `key` {string | Buffer | TypedArray | DataView}
+
+Creates and returns a [Buffer][`Buffer`] object, with the given `algorithm` and
+`key`.
```
Your docs seem to miss information about what the functions actually do.

tniessen commented on doc/api/deprecations.md:
```diff
@@ -634,6 +634,36 @@ Type: Runtime
+
+If the previous key-derivation is required for backward compatiability, the new
+APIs [`crypto.generateLegacyKey()`][] and [`crypto.generateLegacyIV()`][] have
+been added.
```
Not really a causality.

tniessen commented on doc/api/deprecations.md:
```diff
@@ -634,6 +634,36 @@ Type: Runtime
+
+Additionally, for ciphers that require an initialization vector, a proper-length
+initialization vector must be passed to [`crypto.createCipheriv()`][].
+Initialization vectors must never be re-used, especially in modes such as
```
`must` → `should`. The word `must` indicates that calling the function with such parameters leads to undefined behavior or an error, while this is actually just a strong recommendation.

tniessen commented on test/parallel/test-crypto.js:
```diff
@@ -135,6 +135,10 @@ testImmutability(crypto.getCurves);
 // throw, not assert in C++ land.
+if (common.hasFipsCrypto) {
+
+}
```
I assume you are still working on this part.

iangcarroll commented:
Updated. I'll take care of the merge conflicts shortly.

iangcarroll commented:
It's going to be a couple weeks before I can wrap this up. But if there are things that need to be fixed please let me know in the interim.

tniessen commented:
@iangcarroll Let us know if you need any help.

BridgeAR commented:
@iangcarroll do you have the time to pursue this further? :-)

BridgeAR commented:
Ping @iangcarroll

tniessen commented:
I can take over this if necessary. By the way, I would like to reconsider whether we really want to include `generateLegacy*` in core (adding new APIs to support deprecated behavior hardly makes sense).

bnoordhuis commented:
It's not necessary.  [I uploaded a script](https://gist.github.com/bnoordhuis/2de2766d3d3a47ebe41aaaec7e8b14df) last week for computing legacy keys that can be made into a npm module if so desired.

tniessen commented:
> It's not necessary. I uploaded a script last week for computing legacy keys that can be made into a npm module if so desired.

Thanks @bnoordhuis, that's exactly what I had in mind. The crypto API is already quite bloated in my opinion, so I would prefer that very much.

BridgeAR commented:
@tniessen do you still would like to work on this? It would be great if you could open a new PR accordingly. I also think your approach not to add a new API is good.

BridgeAR commented:
Closing due to no further progress.

@iangcarroll thanks a lot for your work! I think it would really be great to get this in and I hope you or someone else might pick this up again.

@tniessen @bnoordhuis I guess one of you two might wants to pick it up?

tniessen commented:
@BridgeAR I will start working on an alternative soon.
